### ISSUE TITLE
"Wizard of Wikipedia" the loss keeps decreasing since the beginning of training

### ISSUE THREAD

**Issue by Ericzh2017** [facebookresearch/ParlAI#1773]
Hello, I have been trying to make incremental improvement on ["Wizard of wikipedia"](https://github.com/facebookresearch/ParlAI/tree/master/projects/wizard_of_wikipedia). Specifically, I have overwritten the "output" method of the modules.
Then I met a wired problem, the loss guadually increased since the beginning of training. What is worse, after a simple change, "loss = -loss", it kept increased instead. 
That is to say, the loss could increase or decrease dependding on what loss was conveyed to the optimizer. However, the loss should be trained to go down gradually given no matter what kind of loss.
Any idea of why this is happening? Should I try to use tensorflow to see whether this problem still exist?

Here was the log when the loss went up:
[ training... ]
[ time:10.0s total_exs:672 epochs:0.01 ] {'exs': 672, 'lr': 2.15e-06, 'num_updates': 21, 'loss': 11.32, 'token_acc': 0.01734, 'nll_loss': 10.61, 'ppl': 40600.0, 'know_loss': nan, 'know_acc': 0.08949, 'know_chance': 0.05994}
[ time:20.0s total_exs:1440 epochs:0.02 ] {'exs': 768, 'lr': 4.55e-06, 'num_updates': 45, 'loss': 13.49, 'token_acc': 0.0194, 'nll_loss': 11.05, 'ppl': 62920.0, 'know_loss': 6.877, 'know_acc': 0.04167, 'know_chance': 0.03858}
[ time:30.0s total_exs:2208 epochs:0.03 ] {'exs': 768, 'lr': 6.949e-06, 'num_updates': 69, 'loss': 14.62, 'token_acc': 0.01711, 'nll_loss': 12.01, 'ppl': 164900.0, 'know_loss': 6.772, 'know_acc': 0.04557, 'know_chance': 0.04002}
[ time:40.0s total_exs:3072 epochs:0.04 ] {'exs': 864, 'lr': 9.649e-06, 'num_updates': 96, 'loss': 16.96, 'token_acc': 0.01795, 'nll_loss': 12.39, 'ppl': 240700.0, 'know_loss': 6.891, 'know_acc': 0.04167, 'know_chance': 0.04029}
[ time:51.0s total_exs:3968 epochs:0.05 ] {'exs': 896, 'lr': 1.245e-05, 'num_updates': 124, 'loss': 17.96, 'token_acc': 0.01874, 'nll_loss': 12.66, 'ppl': 314000.0, 'know_loss': 6.796, 'know_acc': 0.04018, 'know_chance': 0.03923}
[ time:61.0s total_exs:4896 epochs:0.07 ] {'exs': 928, 'lr': 1.535e-05, 'num_updates': 153, 'loss': 19.07, 'token_acc': 0.02049, 'nll_loss': 12.97, 'ppl': 429300.0, 'know_loss': 7.095, 'know_acc': 0.04418, 'know_chance': 0.03954}
[ time:71.0s total_exs:5792 epochs:0.08 ] {'exs': 896, 'lr': 1.815e-05, 'num_updates': 181, 'loss': 18.94, 'token_acc': 0.0188, 'nll_loss': 13.36, 'ppl': 634300.0, 'know_loss': 6.713, 'know_acc': 0.04799, 'know_chance': 0.04043}
[ time:81.0s total_exs:6688 epochs:0.09 ] {'exs': 896, 'lr': 2.095e-05, 'num_updates': 209, 'loss': 19.13, 'token_acc': 0.01958, 'nll_loss': 13.47, 'ppl': 710800.0, 'know_loss': 7.091, 'know_acc': 0.04241, 'know_chance': 0.03997}
[ time:91.0s total_exs:7584 epochs:0.1 ] {'exs': 896, 'lr': 2.375e-05, 'num_updates': 237, 'loss': 19.36, 'token_acc': 0.01928, 'nll_loss': 13.64, 'ppl': 842700.0, 'know_loss': 7.115, 'know_acc': 0.03237, 'know_chance': 0.04028}
[ time:101.0s total_exs:8480 epochs:0.11 ] {'exs': 896, 'lr': 2.655e-05, 'num_updates': 265, 'loss': 19.75, 'token_acc': 0.01948, 'nll_loss': 13.93, 'ppl': 1118000.0, 'know_loss': 7.03, 'know_acc': 0.04688, 'know_chance': 0.03995}
[ time:112.0s total_exs:9408 epochs:0.13 ] {'exs': 928, 'lr': 2.945e-05, 'num_updates': 294, 'loss': 21.12, 'token_acc': 0.01884, 'nll_loss': 14.4, 'ppl': 1790000.0, 'know_loss': 6.858, 'know_acc': 0.04634, 'know_chance': 0.03965}
[ time:122.0s total_exs:10368 epochs:0.14 ] {'exs': 960, 'lr': 3.245e-05, 'num_updates': 324, 'loss': 21.68, 'token_acc': 0.01904, 'nll_loss': 14.27, 'ppl': 1573000.0, 'know_loss': 6.972, 'know_acc': 0.04375, 'know_chance': 0.03936}
[ time:132.0s total_exs:11392 epochs:0.15 ] {'exs': 1024, 'lr': 3.565e-05, 'num_updates': 356, 'loss': 23.32, 'token_acc': 0.02015, 'nll_loss': 14.4, 'ppl': 1789000.0, 'know_loss': 6.826, 'know_acc': 0.05078, 'know_chance': 0.03962}

Added a line "loss = -loss", the loss turned around: 
[ time:10.0s total_exs:672 epochs:0.01 ] {'exs': 672, 'lr': 2.15e-06, 'num_updates': 21, 'loss': -10.85, 'token_acc': 0.01807, 'nll_loss': -10.49, 'ppl': 2.771e-05, 'know_loss': nan, 'know_acc': 0.07244, 'know_chance': 0.06142}
[ time:20.0s total_exs:1504 epochs:0.02 ] {'exs': 832, 'lr': 4.75e-06, 'num_updates': 47, 'loss': -13.41, 'token_acc': 0.01851, 'nll_loss': -10.49, 'ppl': 2.794e-05, 'know_loss': 6.532, 'know_acc': 0.03486, 'know_chance': 0.04005}
[ time:30.0s total_exs:2304 epochs:0.03 ] {'exs': 800, 'lr': 7.249e-06, 'num_updates': 72, 'loss': -13.01, 'token_acc': 0.0134, 'nll_loss': -10.58, 'ppl': 2.549e-05, 'know_loss': 6.85, 'know_acc': 0.04375, 'know_chance': 0.03914}
[ time:40.0s total_exs:3200 epochs:0.04 ] {'exs': 896, 'lr': 1.005e-05, 'num_updates': 100, 'loss': -14.79, 'token_acc': 0.0006249, 'nll_loss': -10.72, 'ppl': 2.209e-05, 'know_loss': 6.089, 'know_acc': 0.03348, 'know_chance': 0.03966}
[ time:50.0s total_exs:4064 epochs:0.05 ] {'exs': 864, 'lr': 1.275e-05, 'num_updates': 127, 'loss': -14.63, 'nll_loss': -11.0, 'ppl': 1.663e-05, 'know_loss': 6.663, 'know_acc': 0.03819, 'know_chance': 0.03865}
[ time:60.0s total_exs:4992 epochs:0.07 ] {'exs': 928, 'lr': 1.565e-05, 'num_updates': 156, 'loss': -16.21, 'nll_loss': -11.34, 'ppl': 1.188e-05, 'know_loss': 6.379, 'know_acc': 0.05927, 'know_chance': 0.03989}
[ time:71.0s total_exs:5888 epochs:0.08 ] {'exs': 896, 'lr': 1.845e-05, 'num_updates': 184, 'loss': -16.02, 'nll_loss': -11.61, 'ppl': 9.043e-06, 'know_loss': 6.325, 'know_acc': 0.02902, 'know_chance': 0.03925}
[ time:81.0s total_exs:6816 epochs:0.09 ] {'exs': 928, 'lr': 2.135e-05, 'num_updates': 213, 'loss': -16.98, 'nll_loss': -11.87, 'ppl': 7.012e-06, 'know_loss': 6.122, 'know_acc': 0.05603, 'know_chance': 0.03985}
[ time:91.0s total_exs:7744 epochs:0.1 ] {'exs': 928, 'lr': 2.425e-05, 'num_updates': 242, 'loss': -17.36, 'nll_loss': -12.12, 'ppl': 5.456e-06, 'know_loss': 5.81, 'know_acc': 0.04418, 'know_chance': 0.03974}
[ time:101.0s total_exs:8704 epochs:0.12 ] {'exs': 960, 'lr': 2.725e-05, 'num_updates': 272, 'loss': -18.38, 'nll_loss': -12.38, 'ppl': 4.186e-06, 'know_loss': 5.109, 'know_acc': 0.04792, 'know_chance': 0.0391}
[ time:111.0s total_exs:9664 epochs:0.13 ] {'exs': 960, 'lr': 3.025e-05, 'num_updates': 302, 'loss': -18.85, 'nll_loss': -12.69, 'ppl': 3.082e-06, 'know_loss': 4.596, 'know_acc': 0.04792, 'know_chance': 0.0395}
[ time:122.0s total_exs:10656 epochs:0.14 ] {'exs': 992, 'lr': 3.335e-05, 'num_updates': 333, 'loss': -20.02, 'nll_loss': -13.03, 'ppl': 2.204e-06, 'know_loss': 4.198, 'know_acc': 0.05544, 'know_chance': 0.03989}
[ time:132.0s total_exs:11648 epochs:0.16 ] {'exs': 992, 'lr': 3.645e-05, 'num_updates': 364, 'loss': -20.57, 'nll_loss': -13.38, 'ppl': 1.545e-06, 'know_loss': 4.125, 'know_acc': 0.04738, 'know_chance': 0.04006}
[ time:142.0s total_exs:12576 epochs:0.17 ] {'exs': 928, 'lr': 3.935e-05, 'num_updates': 393, 'loss': -19.86, 'nll_loss': -13.8, 'ppl': 1.017e-06, 'know_loss': 3.949, 'know_acc': 0.05711, 'know_chance': 0.04032}
[ time:153.0s total_exs:13536 epochs:0.18 ] {'exs': 960, 'lr': 4.235e-05, 'num_updates': 423, 'loss': -21.22, 'nll_loss': -14.25, 'ppl': 6.508e-07, 'know_loss': 3.852, 'know_acc': 0.0625, 'know_chance': 0.04022}
[ time:163.0s total_exs:14528 epochs:0.2 ] {'exs': 992, 'lr': 4.545e-05, 'num_updates': 454, 'loss': -22.71, 'nll_loss': -14.74, 'ppl':

**Comment by stephenroller** [comment]
This usually occurs when there's some sort of numerical instability in the system. Most commonly when the learning rate is too high (but your LR appears quite low). It can also happen if you have something like `X = X / 1e-20` or similar somewhere, or you're using too low precision. The `-loss` aspect isn't really important.

You said you modified output? In what way? 

If you're using fp16, turn it off while you're debugging. If you are using it, also avoid constants bigger than 1e5.

**Comment by Ericzh2017** [comment]
Thanks you. fp16 is set to False. The hp is attached below.

[ optional arguments: ] 
[  knowledge_alpha: 0.95 ]
[ Main ParlAI Arguments: ] 
[  batchsize: 32 ]
[  datapath: /home/cike/ParlAI/data ]
[  datatype: train ]
[  download_path: /home/cike/ParlAI/downloads ]
[  hide_labels: False ]
[  image_mode: raw ]
[  multitask_weights: [1] ]
[  numthreads: 1 ]
[  show_advanced_args: False ]
[  task: wizard_of_wikipedia:generator:random_split ]
[ ParlAI Model Arguments: ] 
[  dict_class: parlai.core.dict:DictionaryAgent ]
[  init_model: None ]
[  model: projects.bg_chat.generator.agents:EndToEndAgent ]
[  model_file: test/model ]
[ Training Loop Arguments: ] 
[  dict_build_first: True ]
[  display_examples: False ]
[  eval_batchsize: None ]
[  evaltask: None ]
[  load_from_checkpoint: False ]
[  max_train_time: -1 ]
[  num_epochs: -1 ]
[  save_after_valid: False ]
[  save_every_n_secs: -1 ]
[  short_final_eval: False ]
[  validation_cutoff: 1.0 ]
[  validation_every_n_epochs: 0.5 ]
[  validation_every_n_secs: -1 ]
[  validation_max_exs: -1 ]
[  validation_metric: ppl ]
[  validation_metric_mode: min ]
[  validation_patience: 12 ]
[  validation_share_agent: False ]
[ Tensorboard Arguments: ] 
[  tensorboard_comment:  ]
[  tensorboard_log: False ]
[  tensorboard_metrics: None ]
[  tensorboard_tag: None ]
[ PytorchData Arguments: ] 
[  batch_length_range: 5 ]
[  batch_sort_cache_type: pop ]
[  batch_sort_field: text ]
[  numworkers: 4 ]
[  pytorch_context_length: -1 ]
[  pytorch_datapath: None ]
[  pytorch_include_labels: True ]
[  pytorch_preprocess: False ]
[  pytorch_teacher_batch_sort: False ]
[  pytorch_teacher_dataset: None ]
[  pytorch_teacher_task: None ]
[  shuffle: False ]
[ Dictionary Loop Arguments: ] 
[  dict_include_test: False ]
[  dict_include_valid: False ]
[  dict_maxexs: -1 ]
[  log_every_n_secs: 10 ]
[ ParlAI Image Preprocessing Arguments: ] 
[  image_cropsize: 224 ]
[  image_size: 256 ]
[ Wizard Dialog Knowledge arguments: ] 
[  include_checked_sentence: True ]
[  include_knowledge: True ]
[  include_knowledge_separator: True ]
[  label_type: response ]
[ GeneratorTeacher Arguments: ] 
[  ignorant_dropout: 0.0 ]
[  only_checked_knowledge: False ]
[ Transformer Arguments: ] 
[  activation: relu ]
[  attention_dropout: 0.0 ]
[  dropout: 0.2 ]
[  embedding_size: 256 ]
[  embeddings_scale: True ]
[  ffn_size: 512 ]
[  learn_positional_embeddings: False ]
[  n_heads: 2 ]
[  n_layers: 5 ]
[  n_positions: 128 ]
[  n_segments: 0 ]
[  relu_dropout: 0.0 ]
[  variant: aiayn ]
[ Torch Generator Agent: ] 
[  beam_block_ngram: 0 ]
[  beam_dot_log: False ]
[  beam_min_length: 1 ]
[  beam_min_n_best: 3 ]
[  beam_size: 1 ]
[  skip_generation: False ]
[ TorchAgent Arguments: ] 
[  add_p1_after_newln: False ]
[  betas: (0.9, 0.98) ]
[  delimiter: 
 ]
[  embedding_projection: random ]
[  embedding_type: fasttext ]
[  fp16: False ]
[  gpu: -1 ]
[  gradient_clip: 0.1 ]
[  history_size: -1 ]
[  label_truncate: None ]
[  learningrate: 0.0005 ]
[  lr_scheduler: invsqrt ]
[  lr_scheduler_decay: 0.5 ]
[  lr_scheduler_patience: 3 ]
[  momentum: 0 ]
[  nesterov: True ]
[  no_cuda: False ]
[  nus: (0.7,) ]
[  optimizer: adam ]
[  person_tokens: False ]
[  rank_candidates: False ]
[  split_lines: False ]
[  text_truncate: None ]
[  truncate: 128 ]
[  update_freq: -1 ]
[  use_reply: label ]
[  warmup_rate: 0.0001 ]
[  warmup_updates: 5000 ]
[ Dictionary Arguments: ] 
[  bpe_debug: False ]
[  dict_endtoken: __end__ ]
[  dict_file: None ]
[  dict_initpath: None ]
[  dict_language: english ]
[  dict_lower: True ]
[  dict_max_ngram_size: -1 ]
[  dict_maxtokens: -1 ]
[  dict_minfreq: 0 ]
[  dict_nulltoken: __null__ ]
[  dict_starttoken: __start__ ]
[  dict_textfields: text,labels,chosen_topic,checked_sentence,knowledge,title ]
[  dict_tokenizer: bpe ]
[  dict_unktoken: __unk__ ]
[ EndToEnd Agent: ] 
[  knowledge_truncate: 32 ]
[  max_knowledge: 32 ]

**Comment by Ericzh2017** [comment]
Is this a overflow or underflow problem?
I make use of the output from transformer decoder to do attention with sigmoid or softmax following. 
I also replace the nn.CrossEntropyLoss() with a pipeline of softmax(), log(), and  nn.NLLLoss() in order to fill in something else.

**Comment by stephenroller** [comment]
If at all possible you should try to use log_softmax instead of softmax followed by a log. The former will be significantly more stable. CrossEntropyLoss is just a wrapper around log_softmax plus NLLLoss.

**Comment by stephenroller** [comment]
It's an out of control gradient problem. Nothing in the options stands out as bad to me, so I would definitely watch for numerically unstable operations in the code you added.

**Comment by Ericzh2017** [comment]
I'm sorry for that the problem behind the code is still under exploring and we need to do much before we manage to solve it. If ever possible shall I email you personally?
It seems that in the code an object of EndToEndAgent uses one dictionary to map between tokens and sub words.
I want to confirm whether the original model uses the same sub-word_to_token and token_to_sub-word dictionary when mapping context and knowledge text to input token sequences and output token sequences to responses text? 
A friend of mine once triggered  a similar bug, because he use two different dictionaries in the encoder and the decoder.

**Comment by stephenroller** [comment]
The dict file released with the original end to end model is correct, and it uses BPE tokens. It looks like you forgot to specify it perhaps though.

**Comment by Ericzh2017** [comment]
I did not specified a dict file path, model.dict.codecs is down load automatically ...

**Comment by stephenroller** [comment]
to be clear, the .dict.codecs file is necessary to perform BPE tokenizations; the .dict file is the actual vocabulary that you will be embedding. You should use the .dict file to look up particular words.

**Comment by Ericzh2017** [comment]
Thanks. Is the .dict file also the vocabulary we are decoding?

**Comment by stephenroller** [comment]
Yes

**Comment by stephenroller** [comment]
Closing this issue for inactivity. Please reopen if you need.
